### ISSUE TITLE
More than 65535 bytes cannot be uploaded over HTTP/2

### ISSUE THREAD

**Issue by manifest** [ninenines/cowboy#1158]
Looks like some kind of issue with window size of HTTP/2 stream.

```bash
$ nghttp -v -H'method:PUT' -H'Content-Type:video/mp4' https://localhost:8443/upload -d example.mp4
[ERROR] Could not connect to the address ::1
Trying next address 127.0.0.1
[  0.014] Connected
[WARNING] Certificate verification failed: unable to verify the first certificate
The negotiated protocol: h2
[  0.071] send SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
[  0.071] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.071] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.071] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.071] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.071] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.071] send HEADERS frame <length=72, flags=0x24, stream_id=13>
          ; END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: POST
          :path: /upload
          :scheme: https
          :authority: localhost:8443
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.22.0
          content-length: 2576890
          method: PUT
          content-type: video/mp4
[  0.071] send DATA frame <length=16384, flags=0x00, stream_id=13>
[  0.071] send DATA frame <length=16384, flags=0x00, stream_id=13>
[  0.071] send DATA frame <length=16384, flags=0x00, stream_id=13>
[  0.071] send DATA frame <length=16383, flags=0x00, stream_id=13>
[  0.077] recv SETTINGS frame <length=0, flags=0x00, stream_id=0>
          (niv=0)
[  0.077] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.078] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
```

```erlang
=INFO REPORT==== 3-May-2017::22:29:02 ===
    request: #{bindings => #{},body_length => 2576890,has_body => true,
               headers =>
                   #{<<"accept">> => <<"*/*">>,
                     <<"accept-encoding">> => <<"gzip, deflate">>,
                     <<"content-length">> => <<"2576890">>,
                     <<"content-type">> => <<"video/mp4">>,
                     <<"method">> => <<"PUT">>,
                     <<"user-agent">> => <<"nghttp2/1.22.0">>},
               host => <<"localhost">>,host_info => undefined,
               method => <<"POST">>,path => <<"/upload">>,
               path_info => undefined,
               peer => {{127,0,0,1},52135},
               pid => <0.200.0>,port => 8443,qs => <<>>,ref => httpd,
               scheme => <<"https">>,streamid => 13,version => 'HTTP/2'}

=INFO REPORT==== 3-May-2017::22:29:17 ===
    {read_body,nofin,65535}

=INFO REPORT==== 3-May-2017::22:29:32 ===
    {read_body,nofin,0}
```

Here is source code of the HTTP handler.

```erlang
-module(httph_upload).

%% HTTP handler callbacks
-export([
  init/2
]).

init(Req, Opts) ->
  error_logger:info_report([{request, Req}]),
  {ok, read_body(Req), Opts}.

read_body(Req0) ->
  case cowboy_req:read_body(Req0) of
    {ok, Data, Req1} ->
      error_logger:info_report([{?FUNCTION_NAME, fin, byte_size(Data)}]),
      Req1;
    {more, Data, Req1} ->
      error_logger:info_report([{?FUNCTION_NAME, nofin, byte_size(Data)}]),
      read_body(Req1)
  end.
```

HTTP/1.1 request works just fine.

```bash
curl -kv --http1.1 -XPUT https://localhost:8443/upload -H 'Content-Type: video/mp4' --data-binary @example.mp4
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 8443 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
* Server certificate: ex.example.org
> PUT /upload HTTP/1.1
> Host: localhost:8443
> User-Agent: curl/7.51.0
> Accept: */*
> Content-Type: video/mp4
> Content-Length: 2576890
> Expect: 100-continue
> 
* Done waiting for 100-continue
* We are completely uploaded and fine
< HTTP/1.1 204 No Content
< 
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
```

```erlang
=INFO REPORT==== 3-May-2017::22:33:20 ===
    request: #{bindings => #{},body_length => 2576890,has_body => true,
               headers =>
                   #{<<"accept">> => <<"*/*">>,
                     <<"content-length">> => <<"2576890">>,
                     <<"content-type">> => <<"video/mp4">>,
                     <<"expect">> => <<"100-continue">>,
                     <<"host">> => <<"localhost:8443">>,
                     <<"user-agent">> => <<"curl/7.51.0">>},
               host => <<"localhost">>,host_info => undefined,
               method => <<"PUT">>,path => <<"/upload">>,
               path_info => undefined,
               peer => {{127,0,0,1},52144},
               pid => <0.200.0>,port => 8443,qs => <<>>,ref => httpd,
               scheme => <<"https">>,streamid => 1,version => 'HTTP/1.1'}

=INFO REPORT==== 3-May-2017::22:33:21 ===
    {read_body,fin,2576890}
```

**Comment by essen** [comment]
Flow control isn't implemented yet, so clients that enforce it won't be able to work with it. This is planned for this month.

**Comment by manifest** [comment]
Good news. Thanks.

**Comment by essen** [comment]
This should be fixed FYI in cad3e3f63d96559f16ee90e661a29063d0b3fc80. Feedback welcome!
